src/main/scala/zio/package.scala

```diff
@@ -17,51 +17,119 @@
 package zio
 
 import java.util.concurrent.ThreadLocalRandom
-import java.util.concurrent.atomic.AtomicReference
 
 import zio.Supervisor.Propagation
 import zio.clock.Clock
 import zio.internal.RingBuffer
 import zio.zmx.diagnostics.{ ZMXConfig, ZMXServer }
 import zio.zmx.diagnostics.graph.{ Edge, Graph, Node }
 import zio.zmx.metrics._
+import zio.stream._
 
 package object zmx extends MetricsDataModel with MetricsConfigDataModel {
 
-  val ZMXSupervisor: Supervisor[Set[Fiber.Runtime[Any, Any]]] =
-    new Supervisor[Set[Fiber.Runtime[Any, Any]]] {
+  object FiberGraph {
+    trait FiberGraph {
```

Comment:
I would move this to the top level instead of in the `FiberGraph` companion object. In fact you can move to a new file. Then in the companion object you can implement a private `apply` method that constructs a `FiberGraph` given a `Graph[Fiber.Runtime[Any, Any], String, String]`. This method can be called from within the `ZMXSupervisor` to clean up the code there. This will also allow you to create your own `FiberGraph` instances by feeding them a `Graph` so you can create your own graph, add a few fibers to it, and verify that you get the expected dumps.

Reply:
thank you very much Adam for the great tip, I moved FiberGraph to separate class. However I am still working on tests, please ignore them for now. I will notify you as soon as I am finished. I am sorry this is taking me so long, my plan is to finish this during the next week.

Reply:
@adamgfraser I believe the PR is ready to review. I changed underlaying AtomicReference of Graph to Ref of a Graph, hopefully that does not break anything. I have written tests on a test graph, that I filled with sample toy fibers (f1 to f8) and following edges:
f1 -> f2
f1 -> f3
f1 -> f4
f4 -> f3
f4 -> f5
f5 -> f6
f7 -> f8
I implemented dump(id, depth) method in a way that it does not include dump of a fiber twice, in case of a cyclical graph. Recursive method remembers what nodes were already traversed. I am not sure if in a running application, this cyclical scenario can occur.....since if "a" supervises "b" and "b" supervises "c", based on what I know I don't expect that it's possible for "c" to supervise "a". Am I right? So In general maybe tree-like structure would be a better fit than the graph to describe fiber supervision relations?

Please let me know if there are other test cases that you would like me to write the tests for. Anyway, I want to thank you very much for your help, this is my first contribution to open source and I am really enjoying the process. I can't wait to start working on next task. If you have any suggestion for a task that is related to this one, please ping me.

Reply:
This looks great! Right now we shouldn't have any cyclic graphs but one discussion has been tracking other information in the graph such as which fibers are waiting on other fibers that could be cyclic, or at least create arrows that point in the "other direction" as supervision. So I think we want to keep the graph structure for now but could definitely revisit at some point if it let us have a more efficient representation!